Re: #1519 Override css and js links

hansthen opened the pull request:
Add a method to the JSCSSMixin that allows users to override a specific link to css or js. Unfortunately this still requires users to know the name of the of the link, which will require looking at the code or introspection.

Conengmo reviewed (changes requested):
Hi Hans, thanks for helping out with this one! Your code looks good, but I want to challenge using another approach here.

Now there's one function with a switch parameter, that has some logic in it to switch based on that parameter. In my opinion this is a pattern that indicates two separate functions would be better suited. Here's an example to illustrate what I mean:

```
def add_js_link(name, url):
     self._add_link(name, url, self.default_js)

def add_css_link(name, url):
     self._add_link(name, url, self.default_css)

def _add_link(name, url, link_array):
    # add the link to the array
```

This way we don't need switching logic. And I think it's a bit more user friendly than having a switching parameter the user must learn how to use.

I'd also opt to not have a separate method to add multiple links. This is easy for users to do themselves, and less code for us to maintain.

To really close the ticket we also need to document this feature in our documentation. I'm thinking maybe a new page in the advanced guide?

hansthen commented:
I updated the code as requested. As to the documentation, adding links for `javascript` and `css` is mostly useful for writing plugins.  Regular use will not typically require this API. What would you think if we added a guide for plugin writers to the advanced documentation and add the documentation there?

Conengmo commented:
> I updated the code as requested.

Thanks!

> As to the documentation, adding links for javascript and css is mostly useful for writing plugins. Regular use will not typically require this API.

I don't think that's true. We've had users ask for this feature so they can make self contained maps that use local resources only. Also, https://github.com/python-visualization/folium/issues/1519 talks about his as a user from a Django context.

For developers we have our contribution guide, which has some guidelines on plugin development. I don't really see the link with this topic here.

I'm also fine with merging this and leaving the ticket open to add documentation in a later stage. What do you prefer?

hansthen commented:
Yeah I see the other use cases now. I think I prefer to add the documentation in this PR.

Conengmo reviewed (approved):
Looks good! Thanks for adding that documentation!
